CRUD-JournalEntries

meb-wtag opened the pull request:
### Description
[JTJ-8 Add Journal](https://welltravel.atlassian.net/jira/software/projects/JTJ/boards/44?selectedIssue=JTJ-8)
Added and edited CRUD for the models 'Journal' ans JournalEntries'. In this branch I started implementing RSpec and i18n aswell. The views stay the same as before.


https://github.com/meb-wtag/journey_me/assets/141820874/a786ce90-faf3-4ecd-b816-b565bc7374b8

meb-wtag committed:
Adding CRUD for entries

Added functionality to add Entries to each Journal. You can delete them aswell.

meb-wtag committed:
Filesystem changed

meb-wtag committed:
Fixxed Bug

Fixxed bug because Id and Journal ID were switched -> Delete and Show function did not work as expected

meb-wtag committed:
Added Sorting

Added sorting in the Journey Table

meb-wtag committed:
deleted unnecessary files

lra-wtag commented on app/controllers/journal_entries_controller.rb:
```diff
@@ -0,0 +1,50 @@
+		params.require(:journal_entry).permit(:title, 
+																					:content)
+	end
+end
```
please correct the indentation

lra-wtag commented on app/models/journal_entry.rb:
```diff
@@ -0,0 +1,4 @@
+class JournalEntry < ApplicationRecord
+
+	belongs_to :journal
+end
```
please correct the indentation

lra-wtag commented on test/fixtures/journal_entries.yml:
```diff
@@ -0,0 +1,11 @@
+# column: value
+#
+two: {}
+# column: value
```
delete this file

lra-wtag commented on test/models/journal_entry_test.rb:
```diff
@@ -0,0 +1,7 @@
+  # test "the truth" do
+  #   assert true
+  # end
+end
```
delete this file

lra-wtag commented on db/schema.rb:
```diff
@@ -10,15 +10,43 @@
     t.datetime "updated_at", null: false
   end
 
+  create_table "users", force: :cascade do |t|
```
why is this change here?

lra-wtag commented on db/migrate/20230904121544_create_journal_entries.rb:
```diff
@@ -0,0 +1,17 @@
+  def up
+    create_table :journal_entries do |t|
+      t.integer "journal_id", :null => false, index: true, foreign_key: true
+      t.integer "user_id"
```
why is the user_id here? I do not see any association with the user model in this PR

meb-wtag committed:
Changes to PR #1

Deleted 1 file, corrected 2 others

meb-wtag committed:
Update journal_entries_controller.rb

lra-wtag reviewed (changes requested):
Lets start writing the model and contoller specs

meb-wtag committed:
All tests so far

Created tests for the Journals-controller and started with the Journal Entries

meb-wtag committed:
Model Tests

Did all tests for validation

meb-wtag committed:
Created all tests

Ready to Merge

chm-wtag commented on app/controllers/journal_entries_controller.rb:
```diff
@@ -0,0 +1,50 @@
+
+  def index
+    @journal_entries = JournalEntry.all
+    redirect_to journals_path
```
do we need this explicit redirection?

meb-wtag replied:
Yes i think so, but i changed it so it fits the logic: 
(it should redirect to the journal, to display all the indexes)      

    @journal_entries = JournalEntry.all
    redirect_to journal_path(@journal)
  end

chm-wtag commented on app/controllers/journal_entries_controller.rb:
```diff
@@ -0,0 +1,50 @@
+class JournalEntriesController < ApplicationController
```
please study about `callbacks` and see if we can use any of them in our code (Hint: `before_action`). Resources:
- [Active Record Callbacks](https://guides.rubyonrails.org/active_record_callbacks.html)
- [Filters](https://guides.rubyonrails.org/action_controller_overview.html#filters )

meb-wtag replied:
I see what yoou mean, but until now i have not implemented a login, so this callbacks are useless, no?

chm-wtag replied:
There are many other use cases! 
`@journal = Journal.find(params[:journal_id])`
This line is repetitive in our current implementation, we can DRY that code using `before_action`.

meb-wtag replied:
Ahhh got it!

chm-wtag commented on app/controllers/journal_entries_controller.rb:
```diff
@@ -0,0 +1,50 @@
+  def destroy
+    @journal = Journal.find(params[:journal_id])
+    @journal_entry = @journal.journal_entries.find(params[:id])
+    @journal_entry.destroy
```
What if destroy operation fails? Please handle that case as well. (Hint: study about `destroy!`)

chm-wtag commented on app/controllers/journal_entries_controller.rb:
```diff
@@ -0,0 +1,50 @@
+  def update
+    @journal = Journal.find(params[:journal_id])
+    @journal_entry = @journal.journal_entries.find(params[:id])
+    @journal_entry = @journal_entry.update(journal_entry_params)
```
What if update operation fails?

chm-wtag commented on app/models/journal.rb:
```diff
@@ -1,6 +1,7 @@
   validates :description, presence: true
-  
+
+  has_many :journal_entries
```
Please study about `dependent_destroy`, `inverse_of`, etc. and let's see if we can use them in our code. Resources:
- [https://guides.rubyonrails.org/association_basics.html](https://guides.rubyonrails.org/association_basics.html)

chm-wtag commented on app/models/journal_entry.rb:
```diff
@@ -0,0 +1,7 @@
+class JournalEntry < ApplicationRecord
+	
+	belongs_to :journal
+
+	validates :title, presence: true
```
2 space indentation

evk-wtag replied:
Indentation should be 2 spaces

chm-wtag commented on db/migrate/20230904121544_create_journal_entries.rb:
```diff
@@ -0,0 +1,16 @@
+class CreateJournalEntries < ActiveRecord::Migration[7.0]
+  def up
+    create_table :journal_entries do |t|
+      t.integer :journal_id, :null => false, index: true, foreign_key: true
```
`:null => false` that's an old format. Let's use the newer formats, `null: false`

chm-wtag reviewed (commented):
- Please delete the .DS_Store files.
- Add dummy data in the seed file.
- Always use 2 space indentation in every file.

Thank you, please reach out in case of any confusions 🎉

evk-wtag commented on app/controllers/journal_entries_controller.rb:
```diff
@@ -0,0 +1,50 @@
+    redirect_to journal_path(@journal) 
+  end
+
+  private
```
Let's add a new line after `private`

evk-wtag commented on app/controllers/journal_entries_controller.rb:
```diff
@@ -0,0 +1,50 @@
+class JournalEntriesController < ApplicationController
+
```
Let's remove this extra line

meb-wtag replied:
why? when do we use a line in between and when not?

evk-wtag replied:
When we are just starting off with a class, we follow this convention of not having empty lines. Unless it is needed for better readability, we will usually avoid redundant lines. You can add a new line in between new methods, or after a list of variables when you are writing a new method. You will be guided in the reviews eventually how we usually maintain it.

Feel free to discuss in this thread if you have any confusions or suggestions..

Thanks for asking!

evk-wtag commented on app/models/journal.rb:
```diff
@@ -1,6 +1,7 @@
 class Journal < ApplicationRecord
-
+  
```
Let's remove this extra line

meb-wtag replied:
When do we use a line in between and when not?

evk-wtag commented on app/views/journals/show.html.erb:
```diff
@@ -9,9 +9,9 @@
-            <th><%= link_to "Title", sort_e: :title%> </th>
-            <th><%= link_to "Updated at", sort_e: :updated_at %> </th>
-            <th><%= link_to "Created at", sort_e: :created_at %> </th>
+            <th><%= link_to "Title", sort: :title%> </th>
```
```suggestion
            <th><%= link_to "Title", sort: :title %> </th>
```

evk-wtag commented on db/migrate/20230904121544_create_journal_entries.rb:
```diff
@@ -0,0 +1,16 @@
+    create_table :journal_entries do |t|
+      t.integer :journal_id, :null => false, index: true, foreign_key: true
+      t.string :title
+      t.text :content
```
Since we have model level validation for title and content to be present, can we also add a db level validation for this?

evk-wtag commented on spec/factories/journals.rb:
```diff
@@ -0,0 +1,6 @@
+FactoryBot.define do
+  factory :journal do
+    title { 'Test' }
+    description { 'abc' }
```
Can we try to integrate and use the gem `ffaker` to generate random texts?

evk-wtag commented on spec/factories/journal_entries.rb:
```diff
@@ -0,0 +1,8 @@
+FactoryBot.define do
+  factory :journal_entry do
+    journal_id {1}
```
Writing only this should work, it is supposed to automatically create a journal when you create a journal_entry
```suggestion
    journal
```

evk-wtag commented on spec/models/journal_entry_spec.rb:
```diff
@@ -0,0 +1,15 @@
+require 'rails_helper'
+RSpec.describe JournalEntry, type: :model do
+
+  journal_test = Journal.new
```
Can you try using FactoryBot.build ?

meb-wtag replied:
Is it necessary to user .build?

I solved it with FactoryBot.create because it is done like that in our sample project aswell

evk-wtag commented:
Could you please add a small PR description regarding your work and a screen-record of the views you are getting? Thanks!

meb-wtag committed:
Corrected PR

meb-wtag commented:
I reviewed all your comments and  corrected and commented where possible. I didn't know where to add a screenshot or more info, so please let me know if you need something...

evk-wtag commented:
> Could you please add a small PR description regarding your work and a screen-record of the views you are getting? Thanks!

Please check this link: https://github.com/meb-wtag/journey_me/pull/3#issue-1880289661 to the PR description. I slightly modified it for you to update it further...

meb-wtag committed:
updated pending spec

chm-wtag commented on app/controllers/journal_entries_controller.rb:
```diff
@@ -0,0 +1,62 @@
+  def index
+    @journal = Journal.find(params[:journal_id])
+    @journal_entries = JournalEntry.all
+    redirect_to journal_path(@journal)
```
do we need these explicit redirections?

meb-wtag replied:
Yes i think so, or is there another way to redirect to the explicit journal?

evk-wtag commented on db/seeds.rb:
```diff
@@ -1,7 +1,2 @@
-#
-#   movies = Movie.create([{ name: "Star Wars" }, { name: "Lord of the Rings" }])
-#   Character.create(name: "Luke", movie: movies.first)
+journal_dummy = Journal.create(title: "Test", description: "text")
+entry_dummy = JournalEntry.create(journal: journal_dummy, title: "Test", content: "text", goal: "abc")
```
Let's use single quotes for the texts

chm-wtag replied:
use loop to populate the db with multiple entries (maybe 5 or 10)

chm-wtag replied:
use `FFaker` for random data generation

evk-wtag commented on spec/controllers/journal_entries_controller_spec.rb:
```diff
@@ -0,0 +1,90 @@
+  end
+
+  let(:valid_params) do
+        FactoryBot.attributes_for(:journal_entry)
+      end
```
Please fix the indentation

evk-wtag commented on spec/controllers/journals_controller_spec.rb:
```diff
@@ -0,0 +1,76 @@
+require 'rails_helper'
+RSpec.describe JournalsController, type: :controller do 	
+	let!(:journal_1) do
+     	FactoryBot.create(:journal)
```
Indentation please

evk-wtag commented on spec/controllers/journals_controller_spec.rb:
```diff
@@ -0,0 +1,76 @@
+        end.to change(Journal, :count).by(1)
+    end
+
+    it 'ridirects to journals_path' do
```
```suggestion
    it 'redirects to journals_path' do
```

evk-wtag commented on spec/factories/journal_entries.rb:
```diff
@@ -0,0 +1,10 @@
+FactoryBot.define do
+  factory :journal_entry do
+    journal
+    title { FFaker::Lorem.sentence }
```
🎉

chm-wtag commented on app/controllers/journals_controller.rb:
```diff
@@ -10,20 +10,31 @@ def new
     @journal = Journal.new(journal_params)
-    if @journal.save
-      redirect_to journals_path 
+      if @journal.save
+      flash[:success] = t('journal.message.success.create')
+      redirect_to journals_path
+    else
+      flash[:error] = t('journal.message.error.create')
+      redirect_to journals_path
     end
```
if @journal.save
      flash[:success] = t('journal.message.success.create')
      redirect_to journals_path
    else
      flash[:error] = t('journal.message.error.create')
      redirect_to journals_path
    end

chm-wtag replied:
indentation in line no 13

chm-wtag commented on app/controllers/journals_controller.rb:
```diff
@@ -10,20 +10,31 @@ def new
-    @journal = Journal.find(params[:id])
+    if !@journal = Journal.find(params[:id])
+      flash[:error] = t('journal.message.error.show')
+    else
```
else what? 😅

chm-wtag commented on app/models/journal_entry.rb:
```diff
@@ -0,0 +1,7 @@
+	belongs_to :journal, inverse_of: :journal_entries
+
+	validates :title, presence: true
+  validates :content, presence: true
```
indentation

evk-wtag commented on spec/models/journal_entry_spec.rb:
```diff
@@ -0,0 +1,21 @@
+    FactoryBot.build(:journal_entry, title: nil)
+  end
+  it 'is not valid without a title' do
+    expect(entry.valid?).to be_falsey
```
Let's introduce the [gem](https://github.com/thoughtbot/shoulda-matchers) 'shoulda-matchers' to make the test cases more concrete.

Let's introduces tests for the model associations and attributes in this PR once you are done learning about this gem.

meb-wtag replied:
Can we please leave this out for now, so i can correct the existing problems?

chm-wtag commented on spec/controllers/journals_controller_spec.rb:
```diff
@@ -0,0 +1,76 @@
+require 'rails_helper'
+RSpec.describe JournalsController, type: :controller do 	
+	let!(:journal_1) do
```
2 space indentation

chm-wtag commented on Gemfile:
```diff
@@ -15,14 +15,19 @@ gem "tzinfo-data", platforms: %i[ mingw mswin x64_mingw jruby ]
 gem "bootsnap", require: false
 gem 'bootstrap', '~> 5.3.1'
 gem 'simple_form'
+gem 'rails-controller-testing'
```
sort the gem names alphabetically, and only use either single or double quotes. We prefer single quotes.

chm-wtag reviewed (commented):
.DS files are still there, please remove them

evk-wtag commented on spec/models/journal_spec.rb:
```diff
@@ -0,0 +1,16 @@
+require 'rails_helper'
+RSpec.describe Journal, type: :model do
+  let!(:journal) do
```
Let's learn about Rspec **describe** blocks and **context** blocks and use them where appropriate.

meb-wtag committed:
Corrected PR

meb-wtag committed:
Update journals_controller.rb

evk-wtag commented on db/seeds.rb:
```diff
@@ -1,7 +1,9 @@
+    content: FFaker::Lorem.sentence,
+    goal: FFaker::Lorem.sentence
+  )
+end
```
Please add a new line at the end of files

evk-wtag commented on spec/controllers/journal_entries_controller_spec.rb:
```diff
@@ -0,0 +1,89 @@
+
+    describe 'DELETE #destroy' do 
+    it 'deletes the journal' do
+        delete :destroy, params: { journal_id: journal.id, id: entry.id }
```
Please fix the indentations. 
You can try using the extension `Rubocop` for easy formatting of .rb files.

chm-wtag commented on app/models/journal_entry.rb:
```diff
@@ -0,0 +1,7 @@
+class JournalEntry < ApplicationRecord
+	
+	belongs_to :journal, inverse_of: :journal_entries
+
+	validates :title, presence: true
+  validates :content, presence: true
+end
```
```
class JournalEntry < ApplicationRecord
  belongs_to :journal, inverse_of: :journal_entries

  validates :title, presence: true
  validates :content, presence: true
end
```

chm-wtag commented on app/controllers/journals_controller.rb:
```diff
@@ -11,15 +10,23 @@ def new
     if @journal.save
-      redirect_to journals_path 
+      flash[:success] = t('journal.message.success.create')
+      redirect_to journals_path
```
let's take this common line out of if-else block

chm-wtag commented on app/controllers/journals_controller.rb:
```diff
@@ -11,15 +10,23 @@ def new
-    @journal.destroy
-
-    redirect_to journals_path
+     if @journal.destroy
```
fix the indentation here

chm-wtag commented on app/controllers/journals_controller.rb:
```diff
@@ -11,15 +10,23 @@ def new
-    redirect_to journals_path
+     if @journal.destroy
+      flash[:success] = t('journal.message.success.delete')
+      redirect_to journals_path
```
no need to put this line in if-else block

chm-wtag commented on config/locales/en.yml:
```diff
@@ -0,0 +1,46 @@
+en:
```
study about i18n-tasks gem and run `i18n-tasks health` [check documentation here](https://github.com/glebm/i18n-tasks#usage)

meb-wtag replied:
thanks. checked it out and played around with it for some time

chm-wtag commented on db/migrate/20230904121544_create_journal_entries.rb:
```diff
@@ -0,0 +1,17 @@
+
+      t.timestamps
+    end 
+    add_index :journal_entries, :title, unique: true
```
you can do this in line no 5 as well.

chm-wtag commented on spec/controllers/journal_entries_controller_spec.rb:
```diff
@@ -0,0 +1,89 @@
+require 'rails_helper'
+RSpec.describe JournalEntriesController, type: :controller do
+  let(:journal) do
+    FactoryBot.create(:journal)
+  end
```
`let(:journal) { FactoryBot.create(:journal) }`

we will use do block only if line limit (110 char) exceeds.

evk-wtag commented on app/controllers/journal_entries_controller.rb:
```diff
@@ -0,0 +1,65 @@
+  def journal_entry_params
+    params.require(:journal_entry).permit(:title, 
+                                          :content)
+  end
```
You can also create a method like
def find_journal_entries
    @journal_entry = @journal.journal_entries.find(params[:id])
end

and use a before_action for this like

before_action :find_journal_entries, only: %i(show destroy)

meb-wtag committed:
Changes on PR

-Implementing shoulda-matchers gem
-format fixes
-before-do actions implemented

meb-wtag committed:
Update journals_controller_spec.rb

evk-wtag commented on app/controllers/journal_entries_controller.rb:
```diff
@@ -0,0 +1,59 @@
+  before_action :set_journal
+  before_action :set_entry, only: %i(show destroy update)
+
+  def set_journal
```
Please rename it to
```suggestion
  def find_journal
```

evk-wtag commented on app/controllers/journal_entries_controller.rb:
```diff
@@ -0,0 +1,59 @@
+    @journal = Journal.find(params[:journal_id])
+  end
+
+  def set_entry
```
```suggestion
  def find_entry
```

evk-wtag commented on app/models/journal_entry.rb:
```diff
@@ -0,0 +1,6 @@
+class JournalEntry < ApplicationRecord	
+  belongs_to :journal, inverse_of: :journal_entries
+
+	validates :title, presence: true
```
```suggestion
  validates :title, presence: true
```

evk-wtag commented on db/migrate/20230904121544_create_journal_entries.rb:
```diff
@@ -0,0 +1,16 @@
+class CreateJournalEntries < ActiveRecord::Migration[7.0]
+  def up
+    create_table :journal_entries do |t|
+      t.integer :journal_id, null: false, index: true, foreign_key: true
+      t.string :title, null: false, unique: true
+      t.text :content, null: false
+      t.text :goal
+
+      t.timestamps
+    end 
+  end
+
+  def down
+    drop_table :journal_entries
+  end
```
You change also just keep one method like:

```suggestion
  def change
    create_table :journal_entries do |t|
      t.integer :journal_id, null: false, index: true, foreign_key: true
      t.string :title, null: false, unique: true
      t.text :content, null: false
      t.text :goal

      t.timestamps
    end 
  end
```

and it will work for both migrating up or down.

meb-wtag replied:
Laila suggested to do it like this

evk-wtag commented on spec/controllers/journals_controller_spec.rb:
```diff
@@ -0,0 +1,68 @@
+require 'rails_helper'
+RSpec.describe JournalsController, type: :controller do 	
+  let(:journal) { FactoryBot.create(:journal) }
+	describe 'GET #index' do 
+		let(:journal2) { FactoryBot.create(:journal) }
+
+   	it 'populates an array of all Journals' do
+			get :index
+    	expect(assigns(:journals)).to match_array [journal, journal2]
+    end
```
Indentation for this file is not correct, you can run Rubocop

chm-wtag commented on app/controllers/journal_entries_controller.rb:
```diff
@@ -0,0 +1,59 @@
+class JournalEntriesController < ApplicationController
+  before_action :set_journal
```
this will be executed before EVERY action. Do we need that? If not, explicitly mention the action names like line no 3 ⬇️

chm-wtag commented on app/controllers/journals_controller.rb:
```diff
@@ -31,3 +36,4 @@ def journal_params
     params.require(:journal).permit(:title, :description)
   end
 end
+  
```
you see a red mark here? It means there's a spacing issue. Please fix it!

meb-wtag replied:
There is no red mark or it fixed itself

chm-wtag commented on app/controllers/journals_controller.rb:
```diff
@@ -31,3 +36,4 @@ def journal_params
     params.require(:journal).permit(:title, :description)
   end
 end
+  
```
```suggestion
  
```

chm-wtag commented on db/seeds.rb:
```diff
@@ -1,7 +1,9 @@
+5.times do
+  journal_dummy = Journal.create(title: FFaker::Lorem.sentence, description: FFaker::Lorem.sentence)
+  journal_entry_dummy = JournalEntry.create(
+    journal: journal,
```
is it `journal_dummy`? 🤔

meb-wtag replied:
Yes, or would you name it differently?

chm-wtag commented on spec/controllers/journal_entries_controller_spec.rb:
```diff
@@ -0,0 +1,73 @@
+
+  describe 'GET #index' do
+    let(:entry2) { FactoryBot.create(:journal_entry, journal: journal) }
+    it 'populates an array of all journal entries' do
+      get :index, params: { journal_id: journal.id, journal_entry: valid_params }
+      expect(assigns(:journal_entries)).to match_array [entry, entry2]
+    end
+
+    it 'redirects to the journals page' do
+      get :index, params: { journal_id: journal.id, journal_entry: valid_params }
+      expect(response).to redirect_to journal_path(journal) 
+    end
```
```suggestion
    get :index, params: { journal_id: journal.id, journal_entry: valid_params }

    it 'populates an array of all journal entries' do
      expect(assigns(:journal_entries)).to match_array [entry, entry2]
    end

    it 'redirects to the journals page' do
      expect(response).to redirect_to journal_path(journal) 
    end
```

chm-wtag replied:
please DRY this in other `describe` blocks as well

meb-wtag replied:
I'd like to leave it like this just to have a reference for both of the options on how to test, if this is ok

chm-wtag commented on spec/models/journal_entry_spec.rb:
```diff
@@ -0,0 +1,11 @@
+require 'rails_helper'
```
add validations for `dependent_destroy` and `inverse_of` as well (wherever applicable)

meb-wtag committed:
Corrected Files and Formated them with Rubocop

lra-wtag commented on app/controllers/journal_entries_controller.rb:
```diff
@@ -0,0 +1,59 @@
+  before_action :find_journal, only: %i[show destroy update]
+  before_action :find_entry, only: %i[show destroy update]
+
+  def find_journal
```
you can put this method inside the private block

lra-wtag commented on app/controllers/journal_entries_controller.rb:
```diff
@@ -0,0 +1,59 @@
+    @journal = Journal.find(params[:journal_id])
+  end
+
+  def find_entry
```
you can put this method inside the private block

lra-wtag commented on app/controllers/journal_entries_controller.rb:
```diff
@@ -0,0 +1,59 @@
+    params.require(:journal_entry).permit(:title,
+                                          :content)
+  end
+end
```
Good work with the i18n :)

meb-wtag replied:
Thanks!

lra-wtag commented on db/schema.rb:
```diff
@@ -10,15 +10,52 @@
+    t.date 'joining_date'
+    t.datetime 'created_at', null: false
+    t.datetime 'updated_at', null: false
+  end
```
why is there the users and tasks here ? I do not see their migrations

meb-wtag replied:
Honestly I don't know. I deleted the 2 Tables out of the file but when I migrate they get added automatically, even if there's no migration-file. Should I leave it like this?

lra-wtag replied:
in such a case go back to your branch where there is a migration and rollback it and make its status as down and come back to this branch and you will see that those migrations will not be here. Please try it.

meb-wtag committed:
Changed

meb-wtag committed:
Update schema.rb

meb-wtag committed:
Update schema.rb

meb-wtag committed:
Update schema.rb
